Working CPU model and few other fixes

pusalieth opened the pull request.

pusalieth commented:
there's a few other fixes, including the requirements.txt to get the python modules up to date, mising imports, etc.

Knucklesfan commented:
Hi, gave this branch a shot since I'm on a mostly AMD machine and liked the prospect of CPU support. On an attempt of loading audio from librispeech, i get this error
`Traceback (most recent call last):
  File "/home/knucklesfan/Documents/faketime/Real-Time-Voice-Cloning-master/toolbox/__init__.py", line 59, in <lambda>
    self.ui.browser_load_button.clicked.connect(lambda: self.load_from_browser())
  File "/home/knucklesfan/Documents/faketime/Real-Time-Voice-Cloning-master/toolbox/__init__.py", line 119, in load_from_browser
    wav = Synthesizer.load_preprocess_wav(fpath)
  File "/home/knucklesfan/Documents/faketime/Real-Time-Voice-Cloning-master/synthesizer/inference.py", line 111, in load_preprocess_wav
    wav = librosa.load(fpath, hparams.sample_rate)[0]
  File "/home/knucklesfan/miniconda3/envs/tensorflow_improved/lib/python3.7/site-packages/librosa/core/audio.py", line 129, in load
    with sf.SoundFile(path) as sf_desc:
  File "/home/knucklesfan/miniconda3/envs/tensorflow_improved/lib/python3.7/site-packages/soundfile.py", line 740, in __init__
    self._file = self._open(file, mode_int, closefd)
  File "/home/knucklesfan/miniconda3/envs/tensorflow_improved/lib/python3.7/site-packages/soundfile.py", line 1263, in _open
    raise TypeError("Invalid file: {0!r}".format(self.name))
TypeError: Invalid file: PosixPath('LibriSpeech/train-clean-100/1034/121119/1034-121119-0026.flac')
` I can't load in any audio from a directory, but the recording feature works as intended.

pusalieth commented:
That last part is your issue.
`TypeError: Invalid file: PosixPath('LibriSpeech/train-clean-100/1034/121119/1034-121119-0026.flac')`
Don't really know what's going on with pathlib, but it's royally screwing up. I was working fine for me, then literally out of no where it stopped working. I'm sure it's some kinda of versioning error.

The issue is here,
`"/home/knucklesfan/Documents/faketime/Real-Time-Voice-Cloning-master/toolbox/__init__.py", line 59`
An fpath object is being passed to the function. How are you running it? I just used the terminal. But whatever is passing the file object to that function, the fpath is f'ed. I personally hate pathlib... adds unnecessary complexity. If you can figure out a way to pass a string directly to the function, it'll work.

Just fyi too, the cpu support is there, however the trained model is for cuda. So, you'll need to train a new checkpoint. There's where I'm currently at.

pusalieth commented:
I'll work on pathlib tonight and see if I can hunt down the issue... All work for no return... ;)

pusalieth commented:
Alright, I think I might have fixed your issue. I'm still working on getting Qt to work with wsl, and a x server... So, you're might gui tester for now.

Knucklesfan commented:
Awesome, giving this a shot right now. I'm on an ubuntu 20.04 machine, and im using conda to actually run the python install. I'll be happy to test anything.

pusalieth commented:
Let me know your results. I'm running all this form wsl in Windows 10 with Ubuntu 20.04 too, just no gui.

adrifcastr commented:
> Let me know your results. I'm running all this form wsl in Windows 10 with Ubuntu 20.04 too, just no gui.
install fails immediately at tensorflow for me
`C:\Users\Adrian\code\Real-Time-Voice-Cloning>pip3.8 install -r requirements.txt
ERROR: Could not find a version that satisfies the requirement tensorflow==1.15 (from -r requirements.txt (line 2)) (fro
m versions: none)
ERROR: No matching distribution found for tensorflow==1.15 (from -r requirements.txt (line 2))`

pusalieth commented:
In your pip usage, make sure you use 3.7. they libraries only work up to python 3.7.x

For tensorflow 1.15.x, you can use pip, or conda. Sometimes anaconda is much easier to control the versions by using specific channels.

Knucklesfan commented:
Okay, the error still occurs with PosixPath, same error format and everything. I checked to see if maybe git didn't update the repo but yeah still on the latest commit and the posixpath still fails.

adrifcastr commented:
> In your pip usage, make sure you use 3.7, the libraries only work up to python 3.7.x. you'll see that as the very first line in requirements.txt.
> 
> For tensorflow 1.15.x, you can use pip, or conda. Sometimes anaconda is much easier to control the versions by using specific channels.

Aight so install worked, but running it errors out at tensorflow,

- Win 10
- Python 3.7.X
- all requirements installed
- latest pytorch installed
- CUDA 10.2
- cudnn installed

Stacktrace: https://hasteb.in/matojeme.coffeescript

pusalieth commented:
I'm getting a virtual machine up and running, so I can work on the gui portion. We'll see where that goes.

The dll issue must be a windows thing. I'm not sure how it loads tensorflow in windows, never used it. Right now, I'm going to debug the qt gui. Next I'll see how hard it is to use windows. In the mean time, if you use wsl like I am, it'll stream line things quite a bit.

adrifcastr commented:
yeah I can use WSL, I'll just need to resetup python and the dependencies first

pusalieth commented:
@adrifcastr Give me a bit. I'm working on windows right now. It's easier than I anticipated. I'll push to the repo soon.

pusalieth commented:
@adrifcastr Alright, I think the issues are fixed with the windows version. I made a .bat file for setup. FYI, pytorch had to be my first install on conda, then with the conda prompt, use pip install for the rest. You may need a new env, or reinstall if you get weird errors. I can confirm though the cpu processing does work on windows.

pusalieth commented:
@Knucklesfan I'm going to start working on your issue and get the gui running.

pusalieth commented:
Alright guys. That should be it. Windows good. GUI good on linux. Just need to train a CPU model now.

I'll probably start hosting some of the files on my own server so the download for models and datasets can be automated as well.

Knucklesfan commented:
Everything is working great! I do now get a NumbaDeprecationWarning, but it's just a warning so I don't think its anything to note. Great work!!

pusalieth commented:
@Knucklesfan Good deal. Good to hear. Yes, the warnings are due to the fact the package uses tensforflow's apiv1, so it'll warning to say change those function calls to update to apiv2. Eventually I'll update them so we can migrate to the latest version.

The only true "error" that I've seen thus far, is something like this.
`Loading checkpoint: synthesizer\saved_models\logs-pretrained\taco_pretrained\tacotron_model.ckpt-278000
2020-05-01 13:08:29.021006: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
2020-05-01 13:08:29.033412: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'nvcuda.dll'; dlerror: nvcuda.dll not found`

That's what the windows error looks like when synthesizing. Linux will say something about failure to load checkpoint. This is due to the fact the checkpoint was produced using cuda, and not a cpu. So there's no model method to recall that checkpoint. I'll have to train a new model, solely on a cpu, and then upload that for usage.

As far as it goes though, the program should all run just fine now on cpu only.

pusalieth commented:
@Knucklesfan Is see the warning now. It's due to librosa calling on numba, when numba is deprecating the import. So, I updated the call on librosa, and submitted a pull request. That'll update librosa for when numba 0.50.0 rolls out.

I also updated this program to use soundfile to write the resultant wav file to disk, since librosa deprecated the write to file function. So, we're now ready for librosa 0.8.0. Whew... LOL

Now I just need to work on tensorflow. I've already implemented a few changes, including while reading the docs to more effective functions for performance.

pusalieth commented:
Alright guys, I think there's really only two things left now. Train new models on a CPU, and replace tf.contrib with tensorflow v2 compatible functions. The second will require a pretty extensive rewrite, so don't expect anything soon. Probably a few weeks at least.

I'll start running the training models on my CPU in the next couple days, and see how long it would take to train models on my CPU.

Knucklesfan commented:
Sounds good. If there's anything you need tested, I'm completely ready to help out.

pusalieth commented:
@Knucklesfan Will do. My only goals left for the software is to get tensorflow 2.0 working. Then I can implement a version for using AMD compute. That'll essentially cover all the bases I think at that point. Nothing really more than could be done. Anything left is just simple maintenance.

Knucklesfan commented:
it works!! I gave it a shot just yesterday and was pleased to hear the AI speaking back to me. It still has a few bugs, though. If you try to load in a paragraph longer than 5-6 lines, it crashes upon attempting to synthesize and vocode, with the core being dumped. I can provide logs in a little bit, but it should be easy enough to reproduce just by opening it up and clicking synthesize and vocode with the default text paragraph.

pusalieth commented:
@Knucklesfan What's your system specs? It might be too low of RAM.

Knucklesfan commented:
16gb DDR4, AMD Ryzen 5 3600x at 3.8ghz and an amd Radeon 5700, that should be good, right?

pusalieth commented:
@Knucklesfan Yeah, that should be more than enough. You might have to send me your error. Open task manager while you run it and see what it does to your RAM and CPU.

Knucklesfan commented:
Memory doesn't ever go above 2.2GiB, and the CPU goes up for a few seconds, and then drops back down after synthesizing is finished. It's a vocoder problem, because the spectrogram is generated, and the crash occurs after that. here's the log in a text file since it's a tad too big to put in a github comment.
[errormessage.txt](https://github.com/CorentinJ/Real-Time-Voice-Cloning/files/4601113/errormessage.txt)

Iamgoofball commented:
Have you made any progress on the tensorflow 2.0 compatibility?

pusalieth commented:
@Iamgoofball 
A little. Many of the calls and variables have been updated, but I believe it was this thread where I stated that nearly the entire program has to be recoded for 2.0 functionality. Compatibility is almost there, but functionality is a whole different ball game. That's an incredible amount of work that won't be done overnight. More like, I'll might slight improvements and take it in bite size chunks over time.

UJJWAL-1711 commented:
for all those attempting to make it on a CPU http://www.xilodyne.com/SBS_RTVC_Demo_Setup 
this thread is amazing easily done in about 30 min but the results are quite flappy not satisfying i don't know why??

matheusfillipe commented:
Is not detecting my audio output devices, this happens if I load a sample and hit play. And there's nothing on the speakers comboBox.

```
Expression 'paInvalidSampleRate' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 2043
Expression 'PaAlsaStreamComponent_InitialConfigure( &self->playback, outParams, self->primeBuffers, hwParamsPlayback, &realSr )' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 2716
Expression 'PaAlsaStream_Configure( stream, inputParameters, outputParameters, sampleRate, framesPerBuffer, &inputLatency, &outputLatency, &hostBufferSizeMode )' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 2837
Traceback (most recent call last):
  File "/home/matheus/programs/Real-time-Voice-fork/toolbox/__init__.py", line 83, in <lambda>
    func = lambda: self.ui.play(self.ui.selected_utterance.wav, Synthesizer.sample_rate)
  File "/home/matheus/programs/Real-time-Voice-fork/toolbox/ui.py", line 142, in play
    sd.play(wav, sample_rate)
  File "/home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/sounddevice.py", line 182, in play
    **kwargs)
  File "/home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/sounddevice.py", line 2498, in start_stream
    **kwargs)
  File "/home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/sounddevice.py", line 1455, in __init__
    **_remove_self(locals()))
  File "/home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/sounddevice.py", line 861, in __init__
    'Error opening {0}'.format(self.__class__.__name__))
  File "/home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/sounddevice.py", line 2653, in _check
    raise PortAudioError(errormsg, err)
sounddevice.PortAudioError: Error opening OutputStream: Invalid sample rate [PaErrorCode -9997]
```

matheusfillipe commented:
Also having this problem when trying to synthesize:
```

Traceback (most recent call last):
  File "/home/matheus/programs/Real-time-Voice-fork/toolbox/__init__.py", line 89, in <lambda>
    func = lambda: self.synthesize() or self.vocode()
  File "/home/matheus/programs/Real-time-Voice-fork/toolbox/__init__.py", line 179, in synthesize
    specs = self.synthesizer.synthesize_spectrograms(texts, embeds)
  File "/home/matheus/programs/Real-time-Voice-fork/synthesizer/inference.py", line 77, in synthesize_spectrograms
    self.load()
  File "/home/matheus/programs/Real-time-Voice-fork/synthesizer/inference.py", line 58, in load
    self._model = Tacotron2(self.checkpoint_fpath, hparams)
  File "/home/matheus/programs/Real-time-Voice-fork/synthesizer/tacotron2.py", line 28, in __init__
    split_infos=split_infos)
  File "/home/matheus/programs/Real-time-Voice-fork/synthesizer/models/tacotron.py", line 146, in initialize
    zoneout=hp.tacotron_zoneout_rate, scope="encoder_LSTM"))
  File "/home/matheus/programs/Real-time-Voice-fork/synthesizer/models/modules.py", line 221, in __init__
    name="encoder_fw_LSTM")
  File "/home/matheus/programs/Real-time-Voice-fork/synthesizer/models/modules.py", line 114, in __init__
    self._cell = tf.contrib.cudnn_rnn.CudnnLSTM(num_units, name=name)
TypeError: __init__() missing 1 required positional argument: 'num_units'

```

matheusfillipe commented:
By doing a print on the sample rate at line 141, that comes from the synthesizer i believe, i have 16000. I hard coded it to pass 44100 to portaudio and the error didn't appear but i didn't hear nothing as well. 

About the other issue with line 114 of "synthesizer/models/modules.py", Adding the named parameter to and adding num_layers to something (kinda random since I don't know nothing about AI :P) Makes it pass this line:

` self._cell = tf.contrib.cudnn_rnn.CudnnLSTM(num_units=num_units, name=name, num_layers=10) `

And get stuck at:
```
    return func(*args, **kwargs)
  File "/home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/tensorflow_core/python/ops/rnn.py", line 449, in bidirectional_dynamic_rnn
    rnn_cell_impl.assert_like_rnncell("cell_fw", cell_fw)
  File "/home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/tensorflow_core/python/ops/rnn_cell_impl.py", line 102, in assert_like_rnncell
    cell_name, cell, ", ".join(errors)))
TypeError: The argument 'cell_fw' (<synthesizer.models.modules.ZoneoutLSTMCell object at 0x7f71e83d3a90>) is not an RNNCell: 'output_size' property is missing, 'state_size' property is missing.
```
I really have the impression I can be running the wrong version of tensorflow but pip says it is 1.15. Maybe is a linux only issue idk?.

In case you are wondering I am running arch linux, I have just updated the nvidia drivers, i have a gtx 1050 ti with 4.2 Gb vram, and I created a virtualenv using conda.

matheusfillipe commented:
I manage to get passed that last error simply by using the else line of the if statement at line 113 of modules.py: 

  ```
  #    if torch.cuda.is_available():
   #        self._cell = tf.contrib.cudnn_rnn.CudnnLSTM(num_units=num_units, name=name, num_layers 5)
   #    else:
      self._cell = tf.contrib.rnn.LSTMBlockCell(num_units=num_units, name=name)
```

Even on tensorflow 1.14  the method CudnnLSTM takes num_layers, which is blank on the pull request. The problem is that It is not going to use the gpu this way right? ... 

This is the full output of the program now:

```
$ python demo_toolbox.py -d /media/matheus/Elements/AI/
/home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/librosa/util/decorators.py:9: NumbaDeprecationWarning: An import was requested from a module that has moved location.
Import requested from: 'numba.decorators', please update to use 'numba.core.decorators' or pin to Numba version 0.48.0. This alias will not be present in Numba version 0.50.0.
  from numba.decorators import jit as optional_jit
/home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/librosa/util/decorators.py:9: NumbaDeprecationWarning: An import was requested from a module that has moved location.
Import of 'jit' requested from: 'numba.decorators', please update to use 'numba.core.decorators' or pin to Numba version 0.48.0. This alias will not be present in Numba version 0.50.0.
  from numba.decorators import jit as optional_jit
Arguments:
    datasets_root:    /media/matheus/Elements/AI
    enc_models_dir:   encoder/saved_models
    syn_models_dir:   synthesizer/saved_models
    voc_models_dir:   vocoder/saved_models
    low_mem:          False

Debug: /home/matheus/projects/TinaVoiceClone/samples/audio_2020-06-05_19-40-01.ogg.wav
Loaded encoder "pretrained.pt" trained to step 1564501
Debug: /home/matheus/projects/TinaVoiceClone/samples/audio_2020-06-05_19-40-07.ogg.wav
Debug: /home/matheus/projects/TinaVoiceClone/samples/audio_2020-06-05_19-40-13.ogg.wav
Debug: /home/matheus/projects/TinaVoiceClone/samples/audio_2020-06-05_19-40-47.ogg.wav
Found synthesizer "pretrained" trained to step 278000
Constructing model: Tacotron
WARNING:tensorflow:From /home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/tensorflow/python/ops/init_ops.py:1251: calling VarianceScaling.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
WARNING:tensorflow:From /home/matheus/programs/Real-time-Voice-fork/synthesizer/models/modules.py:423: conv1d (from tensorflow.python.layers.convolutional) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.keras.layers.Conv1D` instead.
WARNING:tensorflow:From /home/matheus/programs/Real-time-Voice-fork/synthesizer/models/modules.py:424: batch_normalization (from tensorflow.python.layers.normalization) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.BatchNormalization instead.  In particular, `tf.control_dependencies(tf.GraphKeys.UPDATE_OPS)` should not be used (consult the `tf.keras.layers.batch_normalization` documentation).
WARNING:tensorflow:From /home/matheus/programs/Real-time-Voice-fork/synthesizer/models/modules.py:427: dropout (from tensorflow.python.layers.core) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.dropout instead.
WARNING:tensorflow:From /home/matheus/programs/Real-time-Voice-fork/synthesizer/models/modules.py:237: bidirectional_dynamic_rnn (from tensorflow.python.ops.rnn) is deprecated and will be removed in a future version.
Instructions for updating:
Please use `keras.layers.Bidirectional(keras.layers.RNN(cell))`, which is equivalent to this API
WARNING:tensorflow:From /home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/tensorflow/python/ops/rnn.py:464: dynamic_rnn (from tensorflow.python.ops.rnn) is deprecated and will be removed in a future version.
Instructions for updating:
Please use `keras.layers.RNN(cell)`, which is equivalent to this API
WARNING:tensorflow:From /home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/tensorflow/python/ops/rnn.py:244: add_dispatch_support.<locals>.wrapper (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
WARNING:tensorflow:From /home/matheus/programs/Real-time-Voice-fork/synthesizer/models/modules.py:307: MultiRNNCell.__init__ (from tensorflow.python.ops.rnn_cell_impl) is deprecated and will be removed in a future version.
Instructions for updating:
This class is equivalent as tf.keras.layers.StackedRNNCells, and will be replaced by that in Tensorflow 2.0.
WARNING:tensorflow:From /home/matheus/programs/Real-time-Voice-fork/synthesizer/models/modules.py:271: dense (from tensorflow.python.layers.core) is deprecated and will be removed in a future version.
Instructions for updating:
Use keras.layers.dense instead.
WARNING:tensorflow:Entity <bound method ZoneoutLSTMCell.__call__ of <synthesizer.models.modules.ZoneoutLSTMCell object at 0x7f79cc0ba890>> could not be transformed and will be executed as-is. Please report this to the AutgoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: converting <bound method ZoneoutLSTMCell.__call__ of <synthesizer.models.modules.ZoneoutLSTMCell object at 0x7f79cc0ba890>>: ValueError: Failed to parse source code of <bound method ZoneoutLSTMCell.__call__ of <synthesizer.models.modules.ZoneoutLSTMCell object at 0x7f79cc0ba890>>, which Python reported as:
    def __call__(self, inputs, state, scope=None):
        """Runs vanilla LSTM Cell and applies zoneout.
        """
        # Apply vanilla LSTM
        output, new_state = self._cell(inputs, state, scope)

        if self.state_is_tuple:
            (prev_c, prev_h) = state
            (new_c, new_h) = new_state
        else:
            num_proj = self._cell._num_units if self._cell._num_proj is None else \
                                self._cell._num_proj
            prev_c = tf.slice(state, [0, 0], [-1, self._cell._num_units])
            prev_h = tf.slice(state, [0, self._cell._num_units], [-1, num_proj])
            new_c = tf.slice(new_state, [0, 0], [-1, self._cell._num_units])
            new_h = tf.slice(new_state, [0, self._cell._num_units], [-1, num_proj])

        # Apply zoneout
        if self.is_training:
            # nn.dropout takes keep_prob (probability to keep activations) not drop_prob (
                        # probability to mask activations)!
            c = (1 - self._zoneout_cell) * tf.nn.dropout(new_c - prev_c, (1 - self._zoneout_cell)) + prev_c
            h = (1 - self._zoneout_outputs) * tf.nn.dropout(new_h - prev_h, (1 - self._zoneout_outputs)) + prev_h
        else:
            c = (1 - self._zoneout_cell) * new_c + self._zoneout_cell * prev_c
            h = (1 - self._zoneout_outputs) * new_h + self._zoneout_outputs * prev_h

        new_state = tf.compat.v1.nn.rnn_cell.LSTMStateTuple(c, h) if self.state_is_tuple else tf.concat(1, [c,
                                                                                                  h])

        return output, new_state

This may be caused by multiline strings or comments not indented at the same level as the code.
WARNING:tensorflow:Entity <bound method ZoneoutLSTMCell.__call__ of <synthesizer.models.modules.ZoneoutLSTMCell object at 0x7f798c0710d0>> could not be transformed and will be executed as-is. Please report this to the AutgoGraph team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output. Cause: converting <bound method ZoneoutLSTMCell.__call__ of <synthesizer.models.modules.ZoneoutLSTMCell object at 0x7f798c0710d0>>: ValueError: Failed to parse source code of <bound method ZoneoutLSTMCell.__call__ of <synthesizer.models.modules.ZoneoutLSTMCell object at 0x7f798c0710d0>>, which Python reported as:
    def __call__(self, inputs, state, scope=None):
        """Runs vanilla LSTM Cell and applies zoneout.
        """
        # Apply vanilla LSTM
        output, new_state = self._cell(inputs, state, scope)

        if self.state_is_tuple:
            (prev_c, prev_h) = state
            (new_c, new_h) = new_state
        else:
            num_proj = self._cell._num_units if self._cell._num_proj is None else \
                                self._cell._num_proj
            prev_c = tf.slice(state, [0, 0], [-1, self._cell._num_units])
            prev_h = tf.slice(state, [0, self._cell._num_units], [-1, num_proj])
            new_c = tf.slice(new_state, [0, 0], [-1, self._cell._num_units])
            new_h = tf.slice(new_state, [0, self._cell._num_units], [-1, num_proj])

        # Apply zoneout
        if self.is_training:
            # nn.dropout takes keep_prob (probability to keep activations) not drop_prob (
                        # probability to mask activations)!
            c = (1 - self._zoneout_cell) * tf.nn.dropout(new_c - prev_c, (1 - self._zoneout_cell)) + prev_c
            h = (1 - self._zoneout_outputs) * tf.nn.dropout(new_h - prev_h, (1 - self._zoneout_outputs)) + prev_h
        else:
            c = (1 - self._zoneout_cell) * new_c + self._zoneout_cell * prev_c
            h = (1 - self._zoneout_outputs) * new_h + self._zoneout_outputs * prev_h

        new_state = tf.compat.v1.nn.rnn_cell.LSTMStateTuple(c, h) if self.state_is_tuple else tf.concat(1, [c,
                                                                                                  h])

        return output, new_state

This may be caused by multiline strings or comments not indented at the same level as the code.
initialisation done /gpu:0
Initialized Tacotron model. Dimensions (? = dynamic shape):
  Train mode:               False
  Eval mode:                False
  GTA mode:                 False
  Synthesis mode:           True
  Input:                    (?, ?)
  device:                   0
  embedding:                (?, ?, 512)
  enc conv out:             (?, ?, 512)
  encoder out (cond):       (?, ?, 768)
  decoder out:              (?, ?, 80)
  residual out:             (?, ?, 512)
  projected residual out:   (?, ?, 80)
  mel out:                  (?, ?, 80)
  <stop_token> out:         (?, ?)
  Tacotron Parameters       28.439 Million.
Loading checkpoint: synthesizer/saved_models/logs-pretrained/taco_pretrained/tacotron_model.ckpt-278000
2020-06-06 22:18:35.983701: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2020-06-06 22:18:35.988341: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2799925000 Hz
2020-06-06 22:18:35.988811: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x55cfcc7261f0 executing computations on platform Host. Devices:
2020-06-06 22:18:35.988825: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): <undefined>, <undefined>
2020-06-06 22:18:35.989070: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcuda.so.1
2020-06-06 22:18:35.989190: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-06-06 22:18:35.989597: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1640] Found device 0 with properties:
name: GeForce GTX 1050 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.62
pciBusID: 0000:01:00.0
2020-06-06 22:18:35.989635: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudart.so.10.0
2020-06-06 22:18:35.989837: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcublas.so.10.0'; dlerror: libcublas.so.10.0: cannot open shared object file: No such file or directory
2020-06-06 22:18:35.989954: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcufft.so.10.0'; dlerror: libcufft.so.10.0: cannot open shared object file: No such file or directory
2020-06-06 22:18:35.990038: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcurand.so.10.0'; dlerror: libcurand.so.10.0: cannot open shared object file: No such file or directory
2020-06-06 22:18:35.993748: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcusolver.so.10.0'; dlerror: libcusolver.so.10.0: cannot open shared object file: No such file or directory
2020-06-06 22:18:35.993959: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcusparse.so.10.0'; dlerror: libcusparse.so.10.0: cannot open shared object file: No such file or directory
2020-06-06 22:18:36.124646: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7
2020-06-06 22:18:36.124691: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1663] Cannot dlopen some GPU libraries. Skipping registering GPU devices...
2020-06-06 22:18:36.124904: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1181] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-06-06 22:18:36.124954: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1187]      0
2020-06-06 22:18:36.124987: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1200] 0:   N
2020-06-06 22:18:36.128068: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2020-06-06 22:18:36.128683: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x55cfcd793cb0 executing computations on platform CUDA. Devices:
2020-06-06 22:18:36.128718: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): GeForce GTX 1050 Ti, Compute Capability 6.1
2020-06-06 22:18:36.304880: W tensorflow/compiler/jit/mark_for_compilation_pass.cc:1412] (One-time warning): Not using XLA:CPU for cluster because envvar TF_XLA_FLAGS=--tf_xla_cpu_global_jit was not set.  If you want XLA:CPU, either set that envvar, or use experimental_jit_scope to enable XLA:CPU.  To confirm that XLA is active, pass --vmodule=xla_compilation_cache=1 (as a proper command-line flag, not via TF_XLA_FLAGS) or set the envvar XLA_FLAGS=--xla_hlo_profile.
WARNING:tensorflow:From /home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/tensorflow/python/training/saver.py:1276: checkpoint_exists (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
Instructions for updating:
Use standard file APIs to check for files with this prefix.
Building Wave-RNN
Trainable Parameters: 4.481M
Loading model weights at vocoder/saved_models/pretrained/pretrained.pt
DEBUG: 16000
Expression 'paInvalidSampleRate' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 2043
Expression 'PaAlsaStreamComponent_InitialConfigure( &self->playback, outParams, self->primeBuffers, hwParamsPlayback, &realSr )' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 2716
Expression 'PaAlsaStream_Configure( stream, inputParameters, outputParameters, sampleRate, framesPerBuffer, &inputLatency, &outputLatency, &hostBufferSizeMode )' failed in 'src/hostapi/alsa/pa_linux_alsa.c', line: 2837
Traceback (most recent call last):
  File "/home/matheus/programs/Real-time-Voice-fork/toolbox/__init__.py", line 89, in <lambda>
    func = lambda: self.synthesize() or self.vocode()
  File "/home/matheus/programs/Real-time-Voice-fork/toolbox/__init__.py", line 218, in vocode
    self.ui.play(wav, Synthesizer.sample_rate)
  File "/home/matheus/programs/Real-time-Voice-fork/toolbox/ui.py", line 144, in play
    sd.play(wav, sample_rate)
  File "/home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/sounddevice.py", line 182, in play
    **kwargs)
  File "/home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/sounddevice.py", line 2498, in start_stream
    **kwargs)
  File "/home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/sounddevice.py", line 1455, in __init__
    **_remove_self(locals()))
  File "/home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/sounddevice.py", line 861, in __init__
    'Error opening {0}'.format(self.__class__.__name__))
  File "/home/matheus/programs/Real-time-Voice-fork/lib/python3.7/site-packages/sounddevice.py", line 2653, in _check
    raise PortAudioError(errormsg, err)
sounddevice.PortAudioError: Error opening OutputStream: Invalid sample rate [PaErrorCode -9997]

```
Which comes back to the audio playback error... so I think I am very close.

matheusfillipe commented:
Okay, this one was an easy one. Turns out it was trying to use my HDMI output :P which was the device 0 for sounddevice library. I just needed to set the variable as: `sd.default.device = 2` which was my pulse device. 
This could be improved adding a function like:

```
    def get_valid_devices(self, sr):
        supported_devices=[]
        for device in sd.query_devices():
            try:
                sd.check_output_settings(device=device['name'], samplerate=sr)
            except Exception as e:
                print("Sample Rate: ", sr, "Device: ", device['name'], e)
            else:
                supported_devices.append(device['name'])
        return supported_devices

```

And then retrieve the device like `sd.default.device = get_valid_devices(sample_rate)[-1]` Or ideally have a comboBox on the interface for this.

The thing is now it works, but the quality is far from the demo video. I will mess around with it more to see.

CorentinJ commented on toolbox/__init__.py:
```diff
@@ -102,7 +104,11 @@ def load_from_browser(self, fpath=None):
                          self.ui.current_utterance_name)
-            name = str(fpath.relative_to(self.datasets_root))
+
+            if(str(self.datasets_root)[0] == '/' or str(self.datasets_root)[1] == ':'):
+                name = str(fpath.relative_to(self.datasets_root))
+            else:
+                name = os.getcwd() + '/' + str(fpath)
```
Use pathlib

CorentinJ commented on toolbox/__init__.py:
```diff
@@ -111,14 +117,14 @@ def load_from_browser(self, fpath=None):
         else:
-            name = fpath.name
-            speaker_name = fpath.parent.name
+            name = str(fpath).replace('\\', '/')
```
```suggestion
            name = str(fpath).replace("\\", "/")
```

CorentinJ commented on toolbox/__init__.py:
```diff
@@ -111,14 +117,14 @@ def load_from_browser(self, fpath=None):
+        wav = Synthesizer.load_preprocess_wav(name)
         self.ui.log("Loaded %s" % name)
-
+        self.filename = os.path.basename(name)
```
use pathlib

CorentinJ commented on toolbox/__init__.py:
```diff
@@ -211,6 +217,9 @@ def vocoder_progress(i, seq_len, b_size, gen_rate):
         self.ui.play(wav, Synthesizer.sample_rate)
 
+        # Save it
+        sf.write('./Custom_%s.wav' % self.filename, wav, Synthesizer.sample_rate)
```
```suggestion
        sf.write("./Custom_%s.wav" % self.filename, wav, Synthesizer.sample_rate)
```
